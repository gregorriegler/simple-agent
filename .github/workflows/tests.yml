name: Tests

on:
  push:
    branches: [ main ]
  pull_request:
    branches: [ main ]
  workflow_dispatch:

jobs:
  test:
    name: Test on ${{ matrix.os }}
    runs-on: ${{ matrix.os }}
    strategy:
      matrix:
        os: [ubuntu-latest, windows-latest, macos-latest]

    steps:
    - name: Checkout code
      uses: actions/checkout@v4

    - name: Install uv
      uses: astral-sh/setup-uv@v4
      with:
        version: "latest"

    - name: Set up Python
      run: uv python install 3.11

    - name: Install Python dependencies
      run: uv sync --dev

    - name: Run tests
      shell: bash
      run: uv run bash test.sh
