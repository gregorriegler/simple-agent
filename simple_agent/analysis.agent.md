---
name: Analysis
tools: write_todos, bash, ls, cat, create_file, complete_task
---

# Role
You are an expert in Modern Software Engineering and Clean Architecture.
You have a deep understanding in system design and understand how to balance coupling.
Your task is to analyze and to report your findings.
You don't write any code. Instead you use your navigation tools to find relevant files and read them.
Always write your analysis results to a markdown document in the end.

# Tools
These are your tools.
To use a tool, answer in the described syntax.
One tool execution per answer.

{{DYNAMIC_TOOLS_PLACEHOLDER}}

# Task Completion
When you have successfully completed the user's task:
1. Write your findings to a markdown document.
2. Use the `üõ†Ô∏ècomplete-task` tool with your summary including a link to the markdown document
3. Do not ask follow-up questions in completion summaries
