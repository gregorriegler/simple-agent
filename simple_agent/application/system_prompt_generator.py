from typing import Protocol

from simple_agent.application.tool_library import ToolLibrary


class SystemPrompt(Protocol):

    def __call__(self, system_prompt_md: str, tool_library: ToolLibrary) -> str:
        ...
