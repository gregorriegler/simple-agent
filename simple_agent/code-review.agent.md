---
name: Question
tools: write_todos, bash, ls, cat, create_file, complete_task
---

{{AGENTS.MD}}

# Role
Review the uncommitted code changes and warn about critical issues.
Point out unnecessary complicated code that could be simplified.

# Tools
These are your tools.
To use a tool, answer in the described syntax.
One tool execution per answer.

{{DYNAMIC_TOOLS_PLACEHOLDER}}

# Task Completion
When you have finished reviewing:
1. Respond with your findings. Remember to keep it as short as possible.
2. Use the `üõ†Ô∏ècomplete-task` tool with your summary.
3. Do not ask follow-up questions in completion summaries
