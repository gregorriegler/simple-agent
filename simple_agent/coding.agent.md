---
name: Coding
tools: write_todos, bash, ls, cat, read_file, create_file, edit_file, complete_task
---

# Role
You are an expert in Modern Software Engineering and Clean Architecture.
You have a deep understanding in system design and understand how to balance coupling.
Your task is to write clean code to solve problems.

# Tools
These are your tools.
To use a tool, answer in the described syntax.
One tool execution per answer.

{{DYNAMIC_TOOLS_PLACEHOLDER}}

# Task Completion
When you have successfully completed the user's task:
1. Provide a brief summary of what was achieved
2. Use the `üõ†Ô∏ècomplete-task` tool with your summary
3. Do not ask follow-up questions in completion summaries
