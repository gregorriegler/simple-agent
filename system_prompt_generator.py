#!/usr/bin/env python
from tools.tool_library import ToolLibrary

class SystemPromptGenerator:
    """Generates system prompts with dynamically discovered tools"""

    def __init__(self):
        self.tool_library = ToolLibrary()

    def generate_system_prompt(self):
        """Generate the complete system prompt with dynamic tool descriptions"""
        template_content = self._read_system_prompt_template()

        tools_content = self._generate_tools_content()

        return template_content.replace("{{DYNAMIC_TOOLS_PLACEHOLDER}}", tools_content)

    def _read_system_prompt_template(self):
        """Read the system prompt template file"""
        import os
        script_dir = os.path.dirname(os.path.abspath(__file__))
        template_path = os.path.join(script_dir, "system-prompt.md")

        try:
            with open(template_path, 'r', encoding='utf-8') as f:
                return f.read()
        except FileNotFoundError:
            raise FileNotFoundError(f"system-prompt.md template file not found at {template_path}")

    def _generate_tools_content(self):
        tools_lines = []

        for tool in self.tool_library.tools:
            tool_doc = self._generate_tool_documentation(tool)
            if tool_doc:
                tools_lines.append(tool_doc)

        return "\n\n".join(tools_lines)

    def _generate_tool_documentation(self, tool):
        if hasattr(tool, 'get_usage_info'):
            usage_info = tool.get_usage_info()
            return self._format_detailed_tool_doc(usage_info)
        else:
            return self._format_simple_tool_doc(tool)

    def _format_detailed_tool_doc(self, usage_info):
        lines = usage_info.split('\n')
        if not lines:
            return ""

        tool_line = lines[0]
        if not tool_line.startswith('Tool: '):
            return ""

        tool_name = tool_line.replace('Tool: ', '')

        description = ""
        for line in lines[1:]:
            if line.startswith('Description: '):
                description = line.replace('Description: ', '')
                break

        arguments = []
        in_arguments_section = False
        for line in lines:
            if line.strip() == "Arguments:":
                in_arguments_section = True
                continue
            elif in_arguments_section and line.strip().startswith("- "):
                arg = line.strip()[2:]
                arguments.append(arg)
            elif in_arguments_section and line.strip() == "":
                continue
            elif in_arguments_section and not line.strip().startswith("- ") and line.strip():
                break

        doc_lines = [f"## {tool_name}"]
        if description:
            doc_lines.append(f"{description}")
        doc_lines.append("Syntax:")

        if arguments:
            arg_syntax = " ".join([f"<{arg.split(':')[0].strip()}>" for arg in arguments])
            doc_lines.append(f"/{tool_name} {arg_syntax}")
            doc_lines.append("")
            doc_lines.append("Arguments:")
            for arg in arguments:
                doc_lines.append(f"  - {arg}")
        else:
            doc_lines.append(f"/{tool_name} {{arguments}}")

        example = self._extract_example_from_usage_info(usage_info)
        if example:
            doc_lines.append("")
            doc_lines.append("e.g.")
            doc_lines.append(example)

        return "\n".join(doc_lines)

    def _format_simple_tool_doc(self, tool):
        """Format documentation for simple static tools"""
        tool_name = tool.name
        description = getattr(tool, 'description', '')

        doc_lines = [f"## {tool_name}"]
        if description:
            doc_lines.append(description)

        doc_lines.append("Syntax:")

        if tool_name == 'ls':
            doc_lines.append("/ls {path}")
        elif tool_name == 'cat':
            doc_lines.append("/cat {path}")
        elif tool_name == 'test':
            doc_lines.append("/test {directory}")
        elif tool_name == 'revert':
            doc_lines.append("/revert {directory}")
        else:
            doc_lines.append(f"/{tool_name} {{arguments}}")

        return "\n".join(doc_lines)

    def _extract_example_from_usage_info(self, usage_info):
        """Extract example from usage info if available"""
        lines = usage_info.split('\n')
        for i, line in enumerate(lines):
            if 'example' in line.lower() or 'e.g.' in line.lower():
                for j in range(i + 1, len(lines)):
                    if lines[j].strip():
                        return lines[j].strip()
        return None


def main():
    import argparse

    parser = argparse.ArgumentParser(
        description='Generate system prompt with dynamic tool descriptions',
        formatter_class=argparse.RawDescriptionHelpFormatter
    )
    parser.add_argument(
        '--output', '-o',
        help='Output file path (default: stdout)',
        default=None
    )

    args = parser.parse_args()

    generator = SystemPromptGenerator()
    system_prompt = generator.generate_system_prompt()

    if args.output:
        with open(args.output, 'w', encoding='utf-8') as f:
            f.write(system_prompt)
        print(f"System prompt written to {args.output}")
    else:
        print(system_prompt)


if __name__ == "__main__":
    main()
