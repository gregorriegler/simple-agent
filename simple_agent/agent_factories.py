from typing import Callable, Any
from simple_agent.application.agent import Agent
from simple_agent.application.event_bus_protocol import EventBus
from simple_agent.application.input import Input
from simple_agent.application.llm import LLM
from simple_agent.application.session_storage import SessionStorage
from simple_agent.system_prompt_generator import generate_system_prompt


def create_default_agent_factory(
    llm: LLM,
    event_bus: EventBus,
    create_subagent_display: Callable[[str, int], Any],
    create_subagent_input: Callable[[int], Input],
    agent_factory_registry
):
    def factory(
        parent_agent_id: str,
        indent_level: int,
        user_input: Input,
        session_storage: SessionStorage,
        tool_keys=[
            'write_todos',
            'ls',
            'cat',
            'create_file',
            'edit_file',
            'complete_task',
            'bash',
            'subagent'
        ]
    ) -> Agent:
        from simple_agent.tools.all_tools import AllTools

        agent_id = f"{parent_agent_id}/Subagent{indent_level}"
        subagent_tools = AllTools(
            llm,
            indent_level,
            agent_id,
            event_bus,
            user_input,
            create_subagent_display,
            create_subagent_input,
            agent_factory_registry,
            tool_keys
        )
        return Agent(
            agent_id,
            lambda tool_library: generate_system_prompt(subagent_tools),
            subagent_tools,
            llm,
            user_input,
            event_bus,
            session_storage
        )

    return factory
