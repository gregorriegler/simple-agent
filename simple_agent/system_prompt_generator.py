#!/usr/bin/env python
from simple_agent.application.tool_library import ToolLibrary

def extract_tool_keys_from_prompt(system_prompt_md: str) -> list[str]:
    separator = "---"
    if separator not in system_prompt_md:
        return []

    top_section = system_prompt_md.split(separator, 1)[0].strip()
    if not top_section:
        return []

    lines = top_section.split('\n')
    first_line = lines[0].strip()

    if ',' in first_line:
        tool_keys = [key.strip() for key in first_line.split(',') if key.strip()]
        return tool_keys if tool_keys else []

    return []

def generate_system_prompt(system_prompt_md: str, tool_library: ToolLibrary):
    template_content = _read_system_prompt_template(system_prompt_md)
    agents_content = _read_agents_content()
    tools_content = generate_tools_content(tool_library)

    # Replace the dynamic tools placeholder
    result = template_content.replace("{{DYNAMIC_TOOLS_PLACEHOLDER}}", tools_content)

    # Add agents content at the end
    return result + "\n\n" + agents_content



def _read_system_prompt_template(system_prompt_md):
        try:
            from importlib import resources
            content = resources.read_text('simple_agent', '%s' % system_prompt_md)
        except FileNotFoundError:
            import os
            script_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
            template_path = os.path.join(script_dir, "%s" % system_prompt_md)

            try:
                with open(template_path, 'r', encoding='utf-8') as f:
                    content = f.read()
            except FileNotFoundError:
                raise FileNotFoundError(f"%s template file not found at {template_path}" % system_prompt_md)

        return _strip_tool_keys_section(content)


def _strip_tool_keys_section(content: str) -> str:
    separator = "---"
    if separator not in content:
        return content

    parts = content.split(separator, 1)
    if len(parts) < 2:
        return content

    top_section = parts[0].strip()
    if not top_section:
        return content
    return ""


def _read_agents_content():
        import os
        # Look for AGENTS.md in the current working directory
        agents_path = os.path.join(os.getcwd(), "AGENTS.md")

        try:
            # Try UTF-8 first, then fall back to other encodings
            try:
                with open(agents_path, 'r', encoding='utf-8') as f:
                    return f.read()
            except UnicodeDecodeError:
                # Try with UTF-8 and error handling
                with open(agents_path, 'r', encoding='utf-8', errors='replace') as f:
                    return f.read()
        except FileNotFoundError:
            # If AGENTS.md doesn't exist, return empty string to avoid breaking
            return ""

def generate_tools_content(tool_library: ToolLibrary):
    tools_lines = []

    for tool in tool_library.tools:
        tool_doc = _generate_tool_documentation(tool)
        tools_lines.append(tool_doc)

    return "\n\n".join(tools_lines)


def _generate_tool_documentation(tool):
        usage_info = tool.get_usage_info()
        lines = usage_info.split('\n')
        if not lines:
            return ""

        tool_line = lines[0]
        if not tool_line.startswith('Tool: '):
            return ""

        tool_name = tool_line.replace('Tool: ', '')

        description = ""
        for line in lines[1:]:
            if line.startswith('Description: '):
                description = line.replace('Description: ', '')
                break

        arguments = []
        in_arguments_section = False
        for line in lines:
            if line.strip() == "Arguments:":
                in_arguments_section = True
                continue
            elif in_arguments_section and line.strip().startswith("- "):
                arg = line.strip()[2:]
                arguments.append(arg)
            elif in_arguments_section and line.strip() == "":
                continue
            elif in_arguments_section and not line.strip().startswith("- ") and line.strip():
                break

        doc_lines = [f"## {tool_name}"]
        if description:
            doc_lines.append(f"{description}")

        # Simply return the full usage_info after the tool name and description
        remaining_lines = usage_info.split('\n')[2:]  # Skip "Tool: name" and "Description: ..." lines
        if remaining_lines:
            doc_lines.extend(remaining_lines)

        return "\n".join(doc_lines)


def main():
    import argparse
    from simple_agent.tools.all_tools import AllTools

    parser = argparse.ArgumentParser(
        description='Generate system prompt with dynamic tool descriptions',
        formatter_class=argparse.RawDescriptionHelpFormatter
    )
    parser.add_argument(
        '--output', '-o',
        help='Output file path (default: stdout)',
        default=None
    )

    args = parser.parse_args()

    tool_library = AllTools()
    system_prompt = generate_system_prompt('system-prompt.md', tool_library)

    if args.output:
        with open(args.output, 'w', encoding='utf-8') as f:
            f.write(system_prompt)
        print(f"System prompt written to {args.output}")
    else:
        print(system_prompt)


if __name__ == "__main__":
    main()

