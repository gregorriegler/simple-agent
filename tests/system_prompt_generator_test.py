from approvaltests import verify
from unittest.mock import patch

from simple_agent.system_prompt_generator import generate_system_prompt, generate_tools_content
from tests.test_helpers import create_all_tools_for_test


def test_generate_tools_content_only():
    tool_library = create_all_tools_for_test()
    tools_content = generate_tools_content(tool_library)

    verify(tools_content)

def test_generate_orchestrator_agent_system_prompt():
    tool_library = create_all_tools_for_test()
    system_prompt = generate_system_prompt("orchestrator.md", tool_library)
    verify(system_prompt)


def test_generate_complete_system_prompt():
    tool_library = create_all_tools_for_test()

    with patch('simple_agent.system_prompt_generator._read_agents_content') as mock_agents:
        mock_agents.return_value = "# Test AGENTS.md content\nThis is a stub for testing."
        system_prompt = generate_system_prompt('system-prompt.md', tool_library)

    verify(system_prompt)
