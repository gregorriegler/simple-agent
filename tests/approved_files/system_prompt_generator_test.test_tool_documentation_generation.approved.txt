=== ls ===
## ls
List directory contents with detailed information

Arguments:
 - path: string (optional) - Directory path to list (defaults to current directory)

Usage: /ls [path]

Examples:
  /ls
  /ls /tmp/test_path

=== cat ===
## cat
Display file contents with line numbers

Arguments:
 - filename: string (required) - Path to the file to display
 - line_range: string (optional) - Optional line range in format 'start-end' (e.g., '1-10')

Usage: /cat <filename> [line_range]

Examples:
  /cat myfile.txt
  /cat script.py 1-20

=== create-file ===
## create-file
Create new files with optional content

Arguments:
 - filename: string (required) - Path to the file to create (directories will be created automatically)
 - content: string (optional) - Initial content for the file (supports \n for newlines)

Usage: /create-file <filename> [content]

Examples:
  /create-file newfile.txt
  /create-file script.py print("Hello World")
